pilot/pkg/model/gateway.go

```diff
@@ -45,24 +56,101 @@ func MergeGateways(gateways ...Config) *MergedGateway {
 		log.Debugf("MergeGateways: merging gateway %q into %v:\n%v", name, names, gateway)
 		for _, s := range gateway.Servers {
 			log.Debugf("MergeGateways: gateway %q processing server %v", name, s.Hosts)
-			if ss, ok := servers[s.Port.Number]; ok {
-				// TODO: remove this check when Envoy supports filter chain matching so we can expose multiple protocols on the same physical port
-				// ss must have at least one element because the key exists in the map, otherwise we'd be in the else case below.
-				if ss[0].Port.Protocol != s.Port.Protocol {
-					log.Debugf("skipping server: attempting to merge servers for gateway %q into %v but servers have different protocols: want %v have %v",
-						spec.Name, names, ss[0].Port.Protocol, s.Port.Protocol)
-					continue
+			protocol := ParseProtocol(s.Port.Protocol)
+			if gatewayPorts[s.Port.Number] {
+				// We have two servers on the same port. Should we merge?
+				// 1. Yes if both servers are plain text and HTTP
+				// 2. Yes if both servers are using TLS
+				//    if using HTTPS ensure that port name is distinct so that we can setup separate RDS
+				//    for each server (as each server ends up as a separate http connection manager due to filter chain match
+				// 3. No for everything else.
+
+				if p, exists := plaintextServers[s.Port.Number]; exists {
+					currentProto := ParseProtocol(p[0].Port.Protocol)
+					if currentProto != protocol || protocol != ProtocolHTTP {
+						log.Debugf("skipping server on gateway %s port %s.%d.%s: conflict with existing server %s.%d.%s",
+							spec.Name, s.Port.Name, s.Port.Number, s.Port.Protocol, p[0].Port.Name, p[0].Port.Number, p[0].Port.Protocol)
+						continue
+					}
+					rdsName := getRDSRouteName(s)
+					rdsRouteConfigNames[rdsName] = append(rdsRouteConfigNames[rdsName], s)
+				} else {
+					// This has to be a TLS server
+					if isHTTPServer(s) {
```

Comment:
This doesn't match up with the comments: `isHTTPServer` is true for HTTP w/o TLS in addition to HTTPS with TLS config. How are we ensuring this invariant? I assume it's encoded in the route name returned by getRDSRouteName? If so, we should update the comment to make that more obvious.

Reply:
Oh should this be `if isTLSServer(s) {` instead?

Reply:
This is inside the if (portNumber already exists) block. So, a server can either be in plaintext or TLS. Givne that we eliminated plaintext in previous if, this is assuming that the duplicate server is in TLS

Reply:
Doh, yeah sorry I wasn't thinking this through.